docs/hostprovider.md

```diff
@@ -0,0 +1,338 @@
+# Git Credential Manager Core Host Provider
+
+Property|Value
+-|-
+Author(s)|Matthew John Cheetham ([@mjcheetham](https://github.com/mjcheetham))
+Revision|1.0
+Last updated|2020-06-22
+
+## Abstract
+
+Git Credential Manger Core, the cross-platform and cross-host Git credential
+helper, can be extended to support any Git hosting service allowing seemless
+authentication to secured Git repositories by implementing and registering a
+"host provider".
+
+## Table of Contents
+
+- [1. Introduction](#1-introduction)
+  - [1.1. Notational Conventions](#11-notational-conventions)
+  - [1.2. Abbreviations](#12-abbreviations)
+- [2. Implementation](#2-implementation)
+  - [2.1. Registration](#21-registration)
+    - [2.1.2. Ordering](#212-ordering)
+  - [2.2. Handling Requests](#22-handling-requests)
+    - [2.2.1. Rejecting Requests](#221-rejecting-requests)
+  - [2.3. Retrieving Credentials](#23-retrieving-credentials)
+    - [2.3.1 Authentication Prompts](#231-authentication-prompts)
+  - [2.4. Storing Credentials](#24-storing-credentials)
+  - [2.5. Erasing Credentials](#25-erasing-credentials)
+  - [2.6 `HostProvider` base class](#26-hostprovider-base-class)
+    - [2.6.1 `GetCredentialKey`](#261-getcredentialkey)
+    - [2.6.2 `GenerateCredentialAsync`](#262-generatecredentialasync)
+  - [2.7. External Metadata](#27-external-metadata)
+- [3. Helpers](#3-helpers)
+  - [3.1. Discovery](#31-discovery)
+- [4. Error Handling](#4-error-handling)
+
+## 1. Introduction
+
+Git Credential Manager Core (GCM Core) is a host and platform agnostic Git
+credential helper application. Support for authenticating to any Git hosting
+service can be added to GCM Core by creating a custom "host provider" and
+registering it within the product. Host providers can be submitted via a pull
+request on GitHub at <https://github.com/microsoft/Git-Credential-Manager-Core>.
+
+This document outlines the required and expected behaviour of a host provider,
+and what is required to implement and register one.
+
+### 1.1. Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
+specification are to be interpreted as described in
+[[RFC2119](https://tools.ietf.org/html/rfc2119)].
+
+### 1.2. Abbreviations
+
+Throughout this document you may see multiple abbreviations of product names and
+security or credential objects.
+
+"Git Credential Manager Core" is abbreviated to "GCM Core". "Git Credential
+Manager for Windows" is abbreviated to "GCM for Windows", "GCM Windows", or
+"GCM4W". "Git Credential Manager for Mac & Linux" is abbreviated to "GCM for
+Mac/Linux", "GCM Mac/Linux", "GCM4ML", or "Java GCM".
```

Comment:
Could we only use one abbreviation for GCM for Windows and GCM for Mac/Linux?

Reply:
Updated. "GCM [for] Windows" and "GCM [for] Mac/Linux".